Optimize IdP requests to database on login

tack-sap opened the pull request:
This PR addresses Issue #1439

It includes two optimizations w.r.t. retrieving the identity providers from the database:

1) Whenever a login_hint is provided, we do not need to retrieve a full list of all identity providers, but it is sufficient to try to retrieve the identity provider with the given origin. If that does not work an error is returned anyways. So we can optimize the login method in a way, that we only retrieve a single idp, when we have a login_hint. Note that this includes the case, where we click on an account in the account_chooser page.

2) On the pages for the account_chooser and the page for entering the email for idp discovery, the list of all identity providers is retrieved but completely unused. By first determining the parameters and configuration valid for the current request, we can decide if we need to retrieve the identity providers or not. 

As the login method is called several times for each login flow, you will save the unnecessary requests also possibly multiple times per login.  

The optimized scenarios already were covered by Unit tests. The mocks for those were changed, to represent the optimized database requests. But the asserts stay the same, as the visible behavior does not change. 

The visible behavior does not change with this PR - that is also why the Integration Tests do not change. We only optimize the requests to the database (=performance improvement).

If you want to, we can also have a session where we discuss these changes.

tack-sap committed:
Restructure login method to not read all IdentityProviders on login_hint

tack-sap committed:
Move read configurations and parameters to allow earlier decisions

strehle commented:
@aramprice  any concerns?

aramprice commented:
@strehle I'll defer to the UAA team on this. I've been out of the code for too long to offer useful opinions at the moment 😞

bruce-ricard commented:
Hi @tack-sap 

I undid your merge commit and replaced it with a rebase:
```
git reset --hard HEAD~
git rebase develop
git push --force-with-lease
```

I wasn't able to find your commits on the command line because of the merge commit. Could you please try to do rebases in the future? AFAIU merges should only be used when merging a feature/PR branch into the main branch (and even then I personally often like to rebase), and not the other way around.

>The visible behavior does not change with this PR - that is also why the Integration Tests do not change. We only optimize the requests to the database (=performance improvement).

How would you feel about some sort of performance test? We could add a performance test section/directory in our integration tests if it doesn't exist yet. I would like to see a test that shows that indeed in practice this change is useful.

tack-sap commented:
@bruce-ricard Ok, I will use rebase in future - I was just not aware of this rule.

Do you have any preferences on the format of such a performance test? Should I add it here to the PR or are you just interested in having such a test to check manually?

I created a rather simple MockMvcTest today (using StopWatch and looping 1000 times over retrieving the login page with 10 Idps) - with that I see that the new implementation is constantly a few seconds faster compared to the same test executed against the develop branch (on my system: around 8s vs. constantly over 11,5s) - as it saves 10 unnecessary database reads per request. (with a growing number of idps the new version stays at the same time, while the time of the current coding keeps growing with that number).
That's only a few ms per request, but as every browser login does that, is adds up to a big amount (especially retrieving all identity providers over and over again add up to such an amount of DB requests, that the statements showed up in our monitoring as one of the most resource-consuming statements, although they are usually answered very quickly).
Should I add such a test to this PR or do you have something else in mind?

tack-sap committed:
Fix bug with invalid login_hint on login page

tack-sap committed:
Add MockMvc Test to show performance improvement

tack-sap commented:
Fixed a bug, that we found in our internal dev environment 
Added the MockMvc test I mentioned in my last comment that shows the performance improvements

peterhaochen47 commented:
LGTM
